Miopen dialect opt step13 : apply index diff maps in `threadwise_copy` and `threadwise_copy_v2`.

whchung opened the pull request:
This PR use index diff maps for matrix A/B/C, which uses `miopen.threadwise_copy` and `miopen.threadwise_copy_v2` op.

Will revise this PR and implement `F_infinite` algorithm, based on infrastructure built in #231 #230 #229 #228 .

whchung committed:
Apply index diff map in miopen.threadwise_copy_v2 for loads.

whchung committed:
Remove unused codes.

whchung committed:
Apply index diff map in miopen.threadwise_copy_v2 for stores.

whchung committed:
Remove unused codes.

whchung committed:
Factor out common logic.

whchung committed:
Reorder dim_access_order.

Experimental commit.

whchung committed:
Switch between legacy and new approach.

whchung commented:
In another branch `miopen-dialect-opt-step13-exp` where I do more experiments, the current status is:

```
        Current status summary.

        - matrix A
          - global -> VGPR :
            - load  : index diff map
            - store : index diff map
          - VGPR -> LDS:
            - load  : index diff map
            - store : index diff map
        - matrix B
          - global -> VGPR :
            - load  : legacy
            - store : index diff map
          - VGPR -> LDS:
            - load  : index diff map
            - store : index diff map
        - matrix C
          - VGPR -> global:
            - load : index diff map
            - store : legacy
```
It appears logic which has affine maps need to be scrutinized.

whchung commented:
After close scrutiny from @zjing14 and @asroy , existing implementation computing index diff maps needs to be revised and break into smaller PRs. Closing this PR for now.

whchung commented:
Reopen this PR , will revise it and implement `F_infinite` algorithm, based on infrastructure built in #231 #230 #229 #228 .

whchung committed:
Fix clang-format

whchung committed:
Start to use populayeLayerIndices to compute lower-level coordinates.

whchung committed:
Remove composedSource/DestTransform plus renaming some variables.

whchung committed:
Consolidate default lengths of SmallVector instances.

whchung committed:
Make loopIVsPerAccessOrder be an argument rather than a captured object.

whchung committed:
Supply coordinate transformations metadata to index diff map lambda.

whchung committed:
Split source / dest coordinate transformation specifications.

whchung committed:
Start to use the metadata and remove inputType from the lambda.

whchung committed:
Fix unit tests.

whchung committed:
Populate coord transform attributes for source vectors.

whchung committed:
Amend unit tests.

whchung committed:
Fix clang-format.

whchung committed:
Revise computeIndexDiffMap interface.

Output two vectors:
a) lower index diff.
b) lower index updated.

whchung committed:
Move logic to where it's truly needed.

whchung committed:
Start to progressively apply index diff maps.

whchung committed:
Extract common logic to a lambda.

whchung committed:
Remove unused codes.

whchung committed:
Rename some variables.

whchung committed:
Change default lengths of SmallVector instances.

whchung committed:
Carve out lambda from threadwise_copy_v2 to a function.

whchung committed:
Carve out lambda from threadwise_copy_v2 to a function.

populateLayeredIndices ->
  - populateLayeredIndicesWithAffineMap
  - populateLayeredIndicesWithIndexDiffMap

whchung committed:
Populate initial upper and lower indices for index diff map computation.

whchung committed:
Add logic to cope with incomplete metadata.

whchung committed:
Adopt index diff map logic in threadwise_copy. Disabled by default.

whchung committed:
Experimental commit to test 5->3 transfer.

whchung committed:
Change comments.

whchung committed:
Supply UnMerge parameters for matrix C write out logic.

whchung committed:
Supply transform metadata for users of subview op.

whchung committed:
Fix clang-format.

whchung committed:
Proper implementation of index diff maps logic. F_infinite algorithm.

whchung commented:
jenkins: retest this please.

whchung commented:
@asroy this PR is getting closed to completion. I'm kicking off a CI run. Will check with you on 6/1 morning.

whchung commented:
jenkins: retest this please.

whchung commented:
@okakarpa it seems the CI node for `default` is not in a good state.

okakarpa commented:
> @okakarpa it seems the CI node for `default` is not in a good state.
assuming below  is the build error was there any changes in jenkinsfile recently

http://ml-ci.amd.com:21096/job/MLIR/job/MLIR-downstream-unit-test/3458/

okakarpa commented:
> > @okakarpa it seems the CI node for `default` is not in a good state.
> 
> http://ml-ci.amd.com:21096/job/MLIR/job/MLIR-downstream-unit-test/3458/

Looks like there are merge conflicts for the above build

ERROR: Branch not suitable for integration as it does not merge cleanly: Command "git merge --ff bd914d3d2f1492b716537e9b7e8d196c65706472" returned status code 1:
stdout: Auto-merging mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h
CONFLICT (content): Merge conflict in mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h
Automatic merge failed; fix conflicts and then commit the result.

whchung commented:
@okakarpa got it. I’ll fix merge conflicts. It’s introduced by another important PR #227

whchung commented:
need to fix some unit tests. keep `skip-ci` label for now.

asroy commented on mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h:
```diff
@@ -48,15 +48,361 @@ using namespace mlir::miopen;
+        (transformation.getValue() == "Embed")) {
+      auto e = g.get("parameters").template cast<ArrayAttr>();
+      if (transformation.getValue() == "Embed")
+        assert(p.size() == 2);
```
number of upper dimension can be any value for ```embed``` and ```unmerge```

whchung replied:
`UnMerge` can have arbitrary values.
`Embed` is fixated as 2 right now given how `Embed` is used at this moment. We can change later for new use cases.

asroy commented on mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h:
```diff
@@ -48,15 +48,361 @@ using namespace mlir::miopen;
+      auto e = g.get("parameters").template cast<ArrayAttr>();
+      if (transformation.getValue() == "Embed")
+        assert(p.size() == 2);
+      assert(e.size() == p.size());
```
How many parameters does "embed" in MLIR currently have?

For static kernel in current MIOpen, it's N+1.
For dynamic kernel, I removed the last parameter, so it's N

whchung replied:
It's `N` right now.

asroy commented on mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h:
```diff
@@ -48,15 +48,361 @@ using namespace mlir::miopen;
+  // lower level diff map
+  // key : lower level dimension value.
+  // value : lower level diff on that dimension.
+  DenseMap<int64_t, int64_t> lowerIndicesDiffMap;
```
```lowerIndicesDiffMap``` should be ```DenseMap<int64_t, Value>```, because its value may or may not be known at compile time, depending on what is the transformation

One example of ```lowerIndicesDiffMap``` not known at compile time is ```Merge```

whchung replied:
Good finding. I have fixed that in the very latest version.

asroy commented on mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h:
```diff
@@ -48,15 +48,361 @@ using namespace mlir::miopen;
+//===----------------------------------------------------------------------===//
+// Utility function to compute index diff map.
+//===----------------------------------------------------------------------===//
+inline void computeIndexDiffMap(
```
```upperIndicesDiff``` ```lowerIndicesDiff``` should be ```SmallVector<Value, 8>```, because their value may or may not be known at compile time.

whchung replied:
true. Good finding. I have fixed that in the very latest version.

asroy commented on mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h:
```diff
@@ -48,15 +48,361 @@ using namespace mlir::miopen;
+    const SmallVector<AffineMap> &layeredTransform,
+    const SmallVector<SmallVector<Value, 8>, 2> &layeredIndices,
+    const SmallVector<int64_t, 8> &topDiff,
+    SmallVector<SmallVector<int64_t, 8>, 2> &layeredDiffs,
```
should be ```Value```

whchung replied:
Good finding. I have fixed that in the very latest version.

asroy commented on mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h:
```diff
@@ -3428,6 +3768,19 @@ static void affixThreadwiseCopyAttributes(miopen::ThreadwiseCopyOp top,
     top->setAttr("dest_data_per_write", b.getI32IntegerAttr(1));
   }
+  top->setAttr("legacy_load", b.getBoolAttr(true));
+  top->setAttr("legacy_store", b.getBoolAttr(true));
```
These are not buffer-load/store I assume?

whchung replied:
this is a debug switch to switch between:
- the old approach : use `expandAffineMap` to go all the way from the top layer to the bottom layer in one shot.
- the new approach : use `computeIndexDiffMap` to compute index diffs layer by layer

this has nothing to do with load/store instruction selection.

asroy commented on mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h:
```diff
@@ -7038,78 +7600,120 @@ struct ThreadwiseCopyV2RewritePattern
     //   llvm::errs() << sliceLengths[i] << " ";
     // llvm::errs() << "\n";
 
+    // Compute high-level coordinate for dest memref.
```
This is lowering of present value of multi-index of the point of origin? This needs to be done only once when constructing the "thread-copy" object. It's not needed when calling "thread-copy.Run()" every time.

The trick is to build "lower-of-difference" inside "thread_copy.move_slice_window()"

whchung replied:
This is an optimization we can consider in subsequent PRs.

whchung commented on mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h:
```diff
@@ -46,15 +46,416 @@
+  ArrayAttr lowerLayerShape =
+      transformMetadata.get("lower_layer_bounds").template cast<ArrayAttr>();
+
+  // Input:
```
@asroy please help scrutinize comments here against the implementation.

whchung committed:
Fix one unit test.

whchung commented:
For `conv2d_bwd_data` there is another transformation need to consider: `Slice`

whchung committed:
XXX FIXME disable a test for conv2d_bwd_data.

Tame check-mlir for now. This needs to be studied.

whchung committed:
Merge branch 'miopen-dialect' into miopen-dialect-opt-step13

whchung commented on mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h:
```diff
@@ -48,15 +48,416 @@ using namespace mlir::miopen;
+  //       |P| = |Q|
+  //       For each i in P, and its counterpart j in Q
+  //         lower_diff[j] = upper_diff[i]
+  //
```
`Slice` used by `conv2d_bwd_data` needs to be considered here.

whchung replied:
Applied a dirty hack for `conv2d_bwd_data`. `Slice` really builds an identity map which would be optimized away by MLIR by default. Temporarily added a dirty fake map `(d0) -> (d0 + d0 + 1) / 2` to prevent from being optimized away. Need to find a better approach.

+ @asroy / @ltqin / @asleepzzz for awareness.

whchung replied:
Remove all the hacks.

whchung committed:
Fix clang-format.

whchung committed:
Merge branch 'miopen-dialect' into miopen-dialect-opt-step13

whchung committed:
Consider Slice in computeIndexDiffMap. Fix Embed parameters in bwd_data.

whchung committed:
XXX. HACKS for bwd_data.

- Consider Slice in computeIndexDiffMap.
- Fix Embed parameters in bwd_data.
- Populate a fake identity map to prevent it from being optimized.
- Hack a unit test.

whchung committed:
Merge branch 'miopen-dialect' into miopen-dialect-opt-step13

whchung committed:
Revert "XXX FIXME disable a test for conv2d_bwd_data."

This reverts commit 734c407972680fdd10e1d9927162d04ae862bc32.

whchung committed:
Embed affine maps within the metadata of transformations and use them.

Avoid the identity map being optimized away by MLIR when it's embedded as a
part of memref type.

Fix unit tests. Remove those XXX hacks for conv2d_bwd_data.

whchung committed:
Use constantFold whenever possible.

whchung commented:
@okakarpa this particular build is very interesting, it seems something is wrong in `script.sh` ? is it a script used by Jenkins?

http://ml-ci.amd.com:21096/job/MLIR/job/MLIR-downstream-unit-test/3495/console

```
+ [[ ! -f ./compile_commands.json ]]
/home/jenkins/workspace/MLIR/MLIR-downstream-unit-test@tmp/durable-ded6cd3c/script.sh: 2: /home/jenkins/workspace/MLIR/MLIR-downstream-unit-test@tmp/durable-ded6cd3c/script.sh: [[: not found
[Pipeline] sh
+ python3 ./mlir/utils/jenkins/static-checks/premerge-checks.py
clang-format did not modify any files
clang-tidy found error: mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h:16:10: error: 'mlir/Conversion/AffineToStandard/AffineToStandard.h' file not found [clang-diagnostic-error]
clang-tidy found error: mlir/lib/Dialect/MIOpen/Transforms/PassDetail.h:12:10: error: 'mlir/Pass/Pass.h' file not found [clang-diagnostic-error]
clang-tidy found 2 errors and 0 warnings.
```

whchung committed:
Merge branch 'miopen-dialect' into miopen-dialect-opt-step13

whchung commented:
jenkins: retest this please.

whchung committed:
Merge branch 'miopen-dialect' into miopen-dialect-opt-step13
